File: 543.Diameter_of_Binary_Tree.cc
#include <iostream>
#include <cassert>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int findNodeHeights(TreeNode* node, int& maxDiameter) {
        // Base case
        if (node == nullptr) return 0;
        // Recursive case

        int leftHeight = findNodeHeights(node->left, maxDiameter);
        int rightHeight = findNodeHeights(node->right, maxDiameter);

        int currentDiameter = leftHeight + rightHeight;
        maxDiameter = std::max(maxDiameter, currentDiameter);

        return 1 + std::max(leftHeight, rightHeight);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        int maxDiameter = 0;
        // std::cout << "node order: " << root->val << " ";
        findNodeHeights(root, maxDiameter);
        return maxDiameter;
    }
};

/*
    Iterate through each node using pre-order traversal
    For each node, find height of left and right sub-trees
    Compare to currently established max
*/

int main() {
    Solution solution;
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    TreeNode* root2 = new TreeNode(1);
    root2->left = new TreeNode(2);

    TreeNode* root3 = new TreeNode(2);
    root3->left = new TreeNode(3);
    root3->right = new TreeNode(1);

    assert(solution.diameterOfBinaryTree(root) == 3);
    assert(solution.diameterOfBinaryTree(root2) == 1);
    assert(solution.diameterOfBinaryTree(root3) == 2);


    delete root->left->right;
    delete root->left->left;
    delete root->right;
    delete root->left;
    delete root;

    delete root2->left;
    delete root2;

    std::cout << "All tests passed!";
    return 0;
}

File: Python solutions/543. Diameter of Binary Tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxDiameter = 0

        def dfs(node: Optional[TreeNode]) -> int:
            if node is None: return 0
            nonlocal maxDiameter

            leftHeight = dfs(node.left)
            rightHeight = dfs(node.right)

            maxDiameter = max(leftHeight + rightHeight, maxDiameter)
            
            return 1 + max(leftHeight, rightHeight)
        
        dfs(root)
        return maxDiameter

# Helper function to build a tree from a list
def build_tree(values):
    if not values:
        return None
    root = TreeNode(values[0])
    queue = [root]
    front = 0
    index = 1
    while index < len(values):
        node = queue[front]
        front += 1
        if values[index] is not None:
            node.left = TreeNode(values[index])
            queue.append(node.left)
        index += 1
        if index < len(values) and values[index] is not None:
            node.right = TreeNode(values[index])
            queue.append(node.right)
        index += 1
    return root

# Tests
root1 = build_tree([1,2,3,4,5])
assert Solution().diameterOfBinaryTree(root1) == 3, "Test 1 failed!"

root2 = build_tree([1,2])
assert Solution().diameterOfBinaryTree(root2) == 1, "Test 2 failed!"

print("\033[32mAll tests passed!\033[0m\n")